A51-xds-custom-backend-metrics.md

```diff
@@ -0,0 +1,382 @@
+established connection to the backend server to request the load report, and then periodically receiving
+metrics data from the streaming server for each backend.
+The report request should specify metrics reporting interval, the lower bound of which should be capped 
+to a server configurable value, and there is no upper bound unless it is needed in the future.
+
+#### The Server API 
+Provide builtin implementation of `OpenRCAService` streaming service defined by ORCA in each supported language.
+Server applications are free to add such a load reporting server
+in a normal way that they register a service.
+
+The service holds the utilization metrics data in key/value pairs using a hashmap. 
+Expose set-style APIs for user applications to update the metrics data, pseudocode below:
+
+```java
+class OrcaServiceImpl implements OpenRcaService {
+
+// Allow configuring minimum report interval. If not configured or badly configured non-positive, 
+// the default is 30s.
+// In the future, it might be useful to let users specify minimum report interval and default 
+// report interval separately. If the default report interval is not configured or badly configured 
+// non-positive, the default is 1 min. 
+// public OpenRcaServiceImpl(long minInterval, TimeUnit timeUnit, long 
+// defaultInterval, TimeUnit)
+public OrcaServiceImpl(long minReportInterval, TimeUnit unit);
+
+// Update the metrics value corresponding to the specified key.
+void setUtilizationMetric(String key, double value);
+
+// Replace the whole metrics data using the specified map.
+void setAllUtilizationMetrics(Map<String, Double> metrics);
+
+// Remove the metrics data entry corresponding to the specified key.
+void deleteUtilizationMetric(String key);
+
+// Update the CPU utilization metrics data.
+void setCPUUtilizationMetric(double value);
+
+// Clear the CPU utilization metrics data.
+void deleteCPUUtilizationMetric();
+
+// Update the memory utilization metrics data.
+void setMemoryUtilizationMetric(double value);
+
+// Clear the memory utilization metrics data.
+void deleteMemoryUtilizationMetric();
+}
+```
+
+All the methods above should be thread safe. 
+Allowing updating each metric key individually is convenient for a server application to adopt where it 
+comes with multiple isolated components and each generates their own set of metrics.
+
+A minimum reporting interval (30s by default) could be configured when creating the service.
+The minimum report interval is the lower bound of the OOB metrics report period. It means, in the 
+client OOB request message if the `report_interval` 
+goes below the service configured minimum interval then it is treated as the service minimum interval. 
+If `report_interval` is not specified (equal to 0), it is also treated as service minimum.
+There is no upper bound, the largest value you can specify is max integer.
+However, it might be useful to have a separate option called default report interval (default 1 minute) 
```

Comment:
This paragraph can be removed (or, if you really want to keep it, it can be moved to the "Rationale" section to discuss possible future changes).